### ISSUE TITLE
migrate geo, currency into collection for reactive updates 

### ISSUE THREAD

**Issue by aaronjudd** [reactioncommerce/reaction#248]
Changing `dashboard/settings/shop` should change the base currency the shop uses, and then convert what the users sees to their locale currency from the base. This flow seems broken in v0.2.2

**Comment by aaronjudd** [comment]
The flow actually seems to be working now that I'm testing, but requires a refresh on the browser to load the new pricing.  We're not seeing the affect because the pricing quote now has @unblock and thus is skipping getting the new pricing in a reactive manner.

Beyond fixing this issue, I have some ideas on caching these results so that a) the results come from a reactive collection and b) less prone to failed look up c) less api calls (better for any service).

Roughly:
- store results of geocoder, currency in collection
- set defaults in collection
- serves up from collection in future, or else gets from rateUrl
- expires current record 2 hours
- only update access allowed from server

**Comment by aaronjudd** [comment]
Closing we've updated the geo mechanisms, and are also running our own freegeoip server, which was the root of most the performance issues.

**Comment by aaronjudd** [comment]
Re-opened from Gitter discussion https://gitter.im/reactioncommerce/reaction?at=56404293dace6c8c1d1d2ffa 

@newsiberian I think this is what you are describing, PR would be accepted to add the collection/caching with synced-cron as a dependency.

**Comment by newsiberian** [comment]
Hello, @aaronjudd, I'm thinking about autoremove expired rates from the collection, let's say... every 24-48 hours from last update. I was reading about Mongo TTL, but this will clean entire document, we need to clean only 'rate' fields, as I can see. I find a [topic](https://forums.meteor.com/t/solved-remove-old-documents-periodically-good-way-of-doing-it/6853) with similar problem, so as a solution, we can use cron tasks for that. What do you think, maybe there is more gentle way to autoremove fields in document?

Also, I'm worry about this cron package. It is really simple to use, too simple..., but they have a warning on GH page:

>Beware, SyncedCron probably won't work as expected on certain shared hosting providers that shutdown app instances when they aren't receiving requests (like Heroku's free dyno tier or Meteor free galaxy).

so, I find [another one cron package](https://github.com/vsivsi/meteor-job-collection) which is more complex, but at the same time it can be reused in another packages by providing interface to set up some events like promotions, discounts in time, etc and I like it more. May I ask your opinion?

**Comment by mikemurray** [comment]
@newsiberian 

> Beware, SyncedCron probably won't work as expected on certain shared hosting providers that shutdown app instances when they aren't receiving requests (like Heroku's free dyno tier or Meteor free galaxy).

You just need to host any meteor app using synced cron / (or any other NodeJS/Meteor based cron solution) on a server that doesn't auto sleep. Meaning, pay for a tier of Heroku, Meteor galaxy, or host it on Digital Ocean or somewhere else.

If the server is sleeping then crorns will not work; which won't even matter since your app is sleeping too! :smile:

**Comment by aaronjudd** [comment]
We use SyncedCron in another application, and find it works well. It is also the package that I intended to use when I originally added this issue - I still think it's a good choice, but it's really a pretty simple tool to use.

`meteor-job-collection` looks like it provides some very useful features though, I like that there's an actual queue and job status, as well as it's use of a [npm package](https://www.npmjs.com/package/meteor-job) and ddp for remote job management. 

If you want to use `meteor-job-collection` it would be a good choice for other Reaction needs as well.  At first glance, without having actually used the package - I think it's a better way to go.

**Comment by boboci9** [comment]
Hi,
I found an issue in the main.js and in the line 23 I saw a comment `// fix for https://github.com/reactioncommerce/reaction/issues/248` so I figured I should report it here. 
Everything is working fine and then suddenly after some form submits I start getting this error message in the console:
```
TypeError: Cannot read property 'rate' of undefined
    at main.js:30
```
If it's any help I debugged the Locale variable before the if and when the error occurs the Locale object is filled like the Locale.locale is holding all info like { capital, continent, currency, etc} but the Locale.currency is an empty object.

**Comment by aaronjudd** [comment]
@boboci9 thanks for reporting. I created a new issue #600
